pipcompilemulti/features/upgrade.py

```diff
@@ -78,10 +80,19 @@ def reset(self):
         """Clear cached packages."""
         self._env_packages_cache = {}
 
+    @property
+    def package_specs(self):
+        """List of package specs to upgrade."""
+        return self.value or []
+
     @property
     def package_names(self):
         """List of package names to upgrade."""
-        return self.value or []
+        def name_from_spec(name):
+            match = re.match(r'(?P<package>[a-z0-9-_.]+)', name)
+            assert match is not None
```

Comment:
I'm not a fan of having `assert`s outside of tests. The following line will produce an exception anyway.

Reply:
:+1: Alternatively, is there a pattern I can follow to produce a nice error message for the user? Ideally something which helps them understand and likely fix the issue (i.e: that the package name specified likely isn't a valid package).

Reply:
d4ff1ad362c3954027db37d8554e14c0392affe6